AWS: Retain Glue Catalog column comment after updating Iceberg table

lawofcycles opened the pull request:
## What this PR does
This PR addresses the issue where column comments in the Glue catalog are not retained when updating an existing Iceberg table schema. It fixes the problem by modifying the IcebergToGlueConverter.setTableInputInformation method to preserve existing column comments from the Glue table when creating a new TableInput for an update operation.
  
When an existing Glue table is provided, the modified method retrieves the comments from its columns and applies them to the corresponding columns in the new TableInput, but only if the Iceberg table's column does not already have a comment.  
   
This change ensures that any user-defined column comments in the Glue catalog are carried over during table updates, preventing unintentional loss of documentation.
## Testing

Added a new unit test method testSetTableInputInformationWithExistingTable to verify that column comments from an existing Glue table are correctly preserved when creating a new TableInput.  
I'll also add Integration Tests same as [#10199](https://github.com/apache/iceberg/pull/10199).
  
Fixes #10220

lawofcycles commented:
I built the jars and manually tested on Amazon EMR as a Spark runtime.

lawofcycles commented:
Hi @geruh would you review this PR?

geruh commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -234,8 +259,26 @@ static void setTableInputInformation(
+      List<Column> columns = toColumns(metadata);
+      if (existingTable != null) {
+        List<Column> existingColumns = existingTable.storageDescriptor().columns();
+        Map<String, Column> existingColumnMap =
```
What do you think about making this a map of column name to comment here. Since we aren't using any additional information from the existing comments

geruh commented on aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java:
```diff
@@ -501,6 +501,97 @@ public void testColumnCommentsAndParameters() {
   }
 
+  @Test
+  public void testGlueTableColumnCommentsPreserved() {
```
What about the case where we drop a column from a table. I believe today we keep those columns and descriptions. Should we add a test for that?

geruh reviewed (commented):
Looks great, @lawofcycles! I've taken an initial pass. Thanks for the solid work!

lawofcycles commented:
@geruh Thank you for your review! 
I have pushed the changes addressing your review comments.
Could you please take another look when you have a chance?

aajisaka commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -234,8 +258,26 @@ static void setTableInputInformation(
       Optional.ofNullable(properties.get(GLUE_DESCRIPTION_KEY))
           .ifPresent(tableInputBuilder::description);
```
Hi @lawofcycles thank you for your patch!

Now this method has the existing table information, so I think it would be better to move the function to keep the table description https://github.com/aajisaka/iceberg/blob/58c61241bd61de7a4062dbf664691f05eeb2ea53/aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java#L319-L321 into this method.

lawofcycles replied:
Thank you for your review!
I moved keeping the table description logic to IcebergToGlueConverter.

lawofcycles commented:
Hi @geruh would you review the latest change?

aajisaka commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -231,11 +255,33 @@ static void setTableInputInformation(
+        Optional.ofNullable(existingTable.description()).ifPresent(tableInputBuilder::description);
+      }
+
+      List<Column> columns = toColumns(metadata);
```
I'm thinking we can simplify the implementation by creating `existingColumnMap` first and use it inside `addColumnWithDedupe` method. In this way we don't have to replace the new columns.

lawofcycles replied:
I merged existing column comment processing into creating column set on to `toColumns` and `addColumnWithDedupe` method

aajisaka reviewed (approved):
Thank you @lawofcycles

lawofcycles commented:
Hi @amogh-jahagirdar would you review the latest change?

lawofcycles commented:
Hi @amogh-jahagirdar,

I hope this message finds you well. I wanted to kindly follow up on this PR, which addresses an important issue with preserving Glue Catalog column comments when updating Iceberg tables (#10220).

Could you please take a look when you have a moment? Thank you for your time and consideration.

amogh-jahagirdar commented:
Really sorry for the delayed review @lawofcycles , I'll take a look today.

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -231,11 +256,27 @@ static void setTableInputInformation(
+      } else {
+        existingColumnMap = Collections.emptyMap();
+      }
+      List<Column> columns = toColumns(metadata, existingColumnMap);
```
Nit: Newline after the else

lawofcycles replied:
fixed!

amogh-jahagirdar commented on aws/src/test/java/org/apache/iceberg/aws/glue/TestIcebergToGlueConverter.java:
```diff
@@ -311,4 +312,81 @@ public void testSetTableDescription() {
+                    .build())
+            .build();
+
+    Assertions.assertThat(actualTableInput.storageDescriptor().columns())
+        .as("Columns should match")
+        .isEqualTo(expectedTableInput.storageDescriptor().columns());
```
Double checking to make sure this test is strong, `equals` for columns will properly compare against all the fields of the column right? It may be better to explicitly assert the fields we want.

amogh-jahagirdar replied:
Seems like it based on the unit test for testDropColumn, so we should be good.

amogh-jahagirdar commented on aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:
```diff
@@ -219,6 +221,29 @@ static String getTableName(TableIdentifier tableIdentifier, boolean skipNameVali
+  }
+
+  /**
+   * Set Glue table input information based on Iceberg table metadata, optionally preserving
+   * comments from an existing Glue table's columns.
+   *
+   * <p>A best-effort conversion of Iceberg metadata to Glue table is performed to display Iceberg
+   * information in Glue, but such information is only intended for informational human read access
+   * through tools like UI or CLI, and should never be used by any query processing engine to infer
+   * information like schema, partition spec, etc. The source of truth is stored in the actual
+   * Iceberg metadata file defined by the metadata_location table property.
+   *
+   * <p>If an existing Glue table is provided, the comments from its columns will be preserved in
+   * the resulting Glue TableInput. This is useful when updating an existing Glue table to retain
+   * any user-defined comments on the columns.
+   *
+   * @param tableInputBuilder Glue TableInput builder
+   * @param metadata Iceberg table metadata
+   * @param existingTable optional existing Glue table, used to preserve column comments
+   */
```
Wow thank you for this comment! Appreciate that

amogh-jahagirdar reviewed (approved):
@lawofcycles I think this is pretty good, just had some questions and nits. I'll wait a bit if @geruh @rahil-c have any concerns and if not I'll go ahead and merge

rahil-c commented on aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java:
```diff
@@ -221,6 +221,69 @@ public void testUpdateTable() {
+          }
+        });
+
+    table.updateSchema().deleteColumn("c2").deleteColumn("c3").commit();
```
@lawofcycles For my own understanding but when you do the deleteColums and commit shouldn't this sync to catalog and these columns not be present anymore from glue? Or is this commit only changing the iceberg metadata in S3, and thus thats why the assertion is working correctly below as the catalog still has not the updated view of the table and contains the comments?

lawofcycles replied:
@rahil-c  
> Or is this commit only changing the iceberg metadata in S3

Yes.
On current Iceberg's Glue integration, Glue Data Catalog holds all the column info that are registered in Iceberg Table even if these are deleted. IcebergToGlueConverter always try to sync all the historical column to Glue Data Catalog every time schema is updated.
For example, following scenario is expected behaviour.
 
1. Create Table with column x, y, z with Iceberg DDL using Glue Data Catalog as Iceberg Catalog
2. drop column y with Iceberg DDL  -> column y remain on Glue Data Catalog
3. drop column y, z from Glue Data Catalog API
4. add column y2 with Iceberg DDL  ->  column y, z appeared again on Glue Data Catalog in addition to column y2

I think those Information will be used to refer the historical columns on AWS Console when user query tables from such as Athena.

rahil-c replied:
Thanks for the explanation!

Fokko commented:
Let's get this in, thanks @lawofcycles for working on this, and @amogh-jahagirdar and @rahil-c for the review!

lawofcycles commented:
Thanks for reviewing @amogh-jahagirdar , @aajisaka , @geruh and @rahil-c !!

lawofcycles commented:
Thank you for merging @Fokko !
